File: app/src/main/java/ir/radical_app/radical/database/MyDatabase.java
package ir.radical_app.radical.database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;

import androidx.annotation.Nullable;
import ir.radical_app.radical.models.JsonResponse;
import ir.radical_app.radical.models.MessageModel;

public class MyDatabase extends SQLiteOpenHelper {

    private static final String dbName="RadicalDB";
    private static final int dbVersion=1;
    private static final String messageDB = "messages";
    private static final String CategoryDB = "categories";

    private static final String messageDBQuery = "CREATE TABLE messages (id INTEGER primary key autoincrement, sender text, receiver text,message text, title text, date text)";
    private static final String categoriesDBQuery = "CREATE TABLE categories (id INTEGER primary key autoincrement, category_id INTEGER,title text)";

    public MyDatabase(@Nullable Context context) {
        super(context, dbName, null, dbVersion);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL(messageDBQuery);
        db.execSQL(categoriesDBQuery);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
    }


    public void deleteMessageTable(){
        SQLiteDatabase db = getWritableDatabase();
        db.execSQL("DELETE FROM messages");
    }
    public void deleteCategoriesTable(){
        SQLiteDatabase db = getWritableDatabase();
        db.execSQL("DELETE FROM categories");
    }

    public void saveMessage(MessageModel messageModel){
        SQLiteDatabase db = getWritableDatabase();
        ContentValues cv = new ContentValues();
        cv.put("sender",messageModel.getSender());
        cv.put("receiver",messageModel.getReceiver());
        cv.put("message",messageModel.getMessage());
        cv.put("title",messageModel.getTitle());
        cv.put("date",messageModel.getDate());
        db.insert(messageDB,null,cv);
    }
    public void saveCategories(ArrayList<JsonResponse> list){
        SQLiteDatabase db = getWritableDatabase();
        for(int i=0;i<list.size();i++){
            ContentValues cv = new ContentValues();
            JsonResponse jsonResponse = list.get(i);
            cv.put("title",jsonResponse.getName());
            cv.put("category_id",jsonResponse.getCategoryId());
            db.insert(CategoryDB,null,cv);
        }


    }

    public ArrayList<MessageModel> getMessages(){
        ArrayList<MessageModel> list = new ArrayList<>();
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM messages where sender=='مدیر' order by id DESC",null,null);
        if(cursor.moveToFirst()){
            do{
                MessageModel model = new MessageModel();
                model.setMessage(cursor.getString(cursor.getColumnIndex("message")));
                model.setSender(cursor.getString(cursor.getColumnIndex("sender")));
                model.setReceiver(cursor.getString(cursor.getColumnIndex("receiver")));
                model.setTitle(cursor.getString(cursor.getColumnIndex("title")));
                model.setDate(cursor.getString(cursor.getColumnIndex("date")));
                list.add(model);
            }while (cursor.moveToNext());
        }
        return list;
    }
    public ArrayList<MessageModel> getChatMessages(){
        ArrayList<MessageModel> list = new ArrayList<>();
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM messages where sender!='مدیر' order by id ASC",null,null);
        if(cursor.moveToFirst()){
            do{
                MessageModel model = new MessageModel();
                model.setMessage(cursor.getString(cursor.getColumnIndex("message")));
                model.setSender(cursor.getString(cursor.getColumnIndex("sender")));
                model.setReceiver(cursor.getString(cursor.getColumnIndex("receiver")));
                model.setTitle(cursor.getString(cursor.getColumnIndex("title")));
                model.setDate(cursor.getString(cursor.getColumnIndex("date")));
                list.add(model);
            }while (cursor.moveToNext());
        }
        return list;
    }
    public ArrayList<JsonResponse> getCategories(){
        ArrayList<JsonResponse> list = new ArrayList<>();
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM categories",null,null);
        if(cursor.moveToFirst()){
            do{
                JsonResponse model = new JsonResponse();
                model.setName(cursor.getString(cursor.getColumnIndex("title")));
                model.setCategoryId(cursor.getString(cursor.getColumnIndex("category_id")));
                list.add(model);
            }while (cursor.moveToNext());
        }
        return list;
    }
}

File: app/src/main/java/ir/radical_app/radical/database/MyDatabase.kt
package ir.radical_app.radical.database

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import ir.radical_app.radical.models.JsonResponse
import ir.radical_app.radical.models.MessageModel

class MyDatabase constructor(context: Context): SQLiteOpenHelper(context,"RadicalDB",null,1){
    private val messageDB = "messages"
    private val supportDB = "support"
    private val categoryDB = "categories"

    private val messageDBQuery = "CREATE TABLE messages (id INTEGER primary key autoincrement, sender text, receiver text,message text, title text, date text, read INTEGER)"
    private val supportDBQuery = "CREATE TABLE support (id INTEGER primary key autoincrement, sender text, receiver text,message text, title text, date text,read INTEGER)"
    private val categoriesDBQuery = "CREATE TABLE categories (id INTEGER primary key autoincrement, category_id INTEGER,title text)"


    override fun onCreate(db: SQLiteDatabase?) {
      db?.execSQL(messageDBQuery)
      db?.execSQL(supportDBQuery)
      db?.execSQL(categoriesDBQuery)
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
    }

    fun deleteSupportTable(){
        val db = writableDatabase
        db.execSQL("DELETE FROM $supportDB")
    }
    fun updateSupportDB(){
        val db = writableDatabase
        db.execSQL("update $supportDB set read=1 where read=0")
    }

    fun deleteMessageTable(){
        val db = writableDatabase
        db.execSQL("DELETE FROM $messageDB")
    }
    fun readMessage(id:String){
        val db=writableDatabase
        db.execSQL("update $messageDB set read=1 where id=$id")
    }

    fun deleteCategoriesTable(){
        val db = writableDatabase
        db.execSQL("DELETE FROM $categoryDB")
    }

    fun saveMessage(messageModel:MessageModel){
        val db = writableDatabase
        val cv = ContentValues()
        cv.put("sender",messageModel.sender);
        cv.put("receiver",messageModel.receiver);
        cv.put("message",messageModel.message);
        cv.put("title",messageModel.title);
        cv.put("date",messageModel.date);
        cv.put("read",0);
        db.insert(messageDB,null,cv)
    }

    fun saveSupport(messageModel:MessageModel){
        val db = writableDatabase
        val cv = ContentValues()
        cv.put("sender",messageModel.sender);
        cv.put("receiver",messageModel.receiver);
        cv.put("message",messageModel.message);
        cv.put("title",messageModel.title);
        cv.put("date",messageModel.date);
        cv.put("read",messageModel.read);
        db.insert(supportDB,null,cv)
    }

    fun saveCategories(list : ArrayList<JsonResponse>){
        val db = writableDatabase
        for(i in 0 until list.size){
            val cv = ContentValues()
            val jsonResponse = list[i]
            cv.put("title",jsonResponse.name);
            cv.put("category_id",jsonResponse.categoryId);
            db.insert(categoryDB,null,cv)
        }
    }

    fun getMessages():ArrayList<MessageModel>{
        val list = ArrayList<MessageModel>()
        val db = readableDatabase
        val cursor:Cursor=db.rawQuery("SELECT * FROM $messageDB order by id DESC",null,null)
        if(cursor.moveToFirst()){
            do{
                val model =MessageModel()
                model.id=cursor.getInt(cursor.getColumnIndex("id")).toString()
                model.read=cursor.getInt(cursor.getColumnIndex("read")).toString()
                model.message=cursor.getString(cursor.getColumnIndex("message"))
                model.sender=cursor.getString(cursor.getColumnIndex("sender"))
                model.receiver=cursor.getString(cursor.getColumnIndex("receiver"))
                model.title=cursor.getString(cursor.getColumnIndex("title"))
                model.date=cursor.getString(cursor.getColumnIndex("date"))
                list.add(model)
            }while (cursor.moveToNext())
        }
        cursor.close()
        return list
    }

    fun getReadMessages():Int{
        val db = readableDatabase
        val cursor:Cursor=db.rawQuery("SELECT COUNT(*) AS count FROM $messageDB where read=0;",null,null)
        return if(cursor.moveToFirst()) cursor.getInt(cursor.getColumnIndex("count")) else 0
    }

    fun getChatMessages():ArrayList<MessageModel>{
        val list = ArrayList<MessageModel>()
        val db = readableDatabase
        val cursor:Cursor=db.rawQuery("SELECT * FROM $supportDB order by id ASC",null,null)
        if(cursor.moveToFirst()){
            do{
                val model=MessageModel()
                model.message=cursor.getString(cursor.getColumnIndex("message"))
                model.sender=cursor.getString(cursor.getColumnIndex("sender"))
                model.receiver=cursor.getString(cursor.getColumnIndex("receiver"))
                model.title=cursor.getString(cursor.getColumnIndex("title"))
                model.date=cursor.getString(cursor.getColumnIndex("date"))
                list.add(model)
            }while (cursor.moveToNext())
        }
        cursor.close()
        updateSupportDB()
        return list
    }

    fun getReadSupport():Int{
        val db = readableDatabase
        val cursor:Cursor=db.rawQuery("SELECT COUNT(*) AS count FROM $supportDB where read=0;",null,null)
        return if(cursor.moveToFirst()) cursor.getInt(cursor.getColumnIndex("count")) else 0
    }

    fun getCategories():ArrayList<JsonResponse>{
        val list = ArrayList<JsonResponse>()
        val db = readableDatabase
        val cursor:Cursor=db.rawQuery("SELECT * FROM $categoryDB",null,null)
        if(cursor.moveToFirst()){
            do{
                val model= JsonResponse()
                model.name = cursor.getString(cursor.getColumnIndex("title"))
                model.categoryId = cursor.getString(cursor.getColumnIndex("category_id"))
                list.add(model)

            }while (cursor.moveToNext())

        }
        cursor.close()
        return list
    }

}